File: ios/MultiplatformBleAdapter/BleAdapter.swift
import Foundation

@objc public protocol BleAdapter {

    // MARK: - Properties

    var delegate: BleClientManagerDelegate? { get set }

    // MARK: - Initializers

    init(queue: DispatchQueue, restoreIdentifierKey: String?)
    func invalidate()

    // MARK: - Functions

    func cancelTransaction(_ transactionId: String)

    func setLogLevel(_ logLevel: String)

    func logLevel(_ resolve: Resolve,
                  reject: Reject)

    func enable(_ transactionId: String,
                resolve: Resolve,
                reject: Reject)

    func disable(_ transactionId: String,
                 resolve: Resolve,
                 reject: Reject)

    func state(_ resolve: Resolve,
               reject: Reject)

    func startDeviceScan(_ filteredUUIDs: [String]?,
                         options:[String:AnyObject]?)

    func stopDeviceScan()

    func readRSSIForDevice(_ deviceIdentifier: String,
                           transactionId: String,
                           resolve: @escaping Resolve,
                           reject: @escaping Reject)

    func requestMTUForDevice(_ deviceIdentifier: String,
                             mtu: Int,
                             transactionId: String,
                             resolve: @escaping Resolve,
                             reject: @escaping Reject)

    func requestConnectionPriorityForDevice(_ deviceIdentifier: String,
                                            connectionPriority: Int,
                                            transactionId: String,
                                            resolve: @escaping Resolve,
                                            reject: @escaping Reject)

    func devices(_ deviceIdentifiers: [String],
                 resolve: @escaping Resolve,
                 reject: @escaping Reject)

    func connectedDevices(_ serviceUUIDs: [String],
                          resolve: @escaping Resolve,
                          reject: @escaping Reject)

    func connectToDevice(_ deviceIdentifier: String,
                         options:[String: AnyObject]?,
                         resolve: @escaping Resolve,
                         reject: @escaping Reject)

    func cancelDeviceConnection(_ deviceIdentifier: String,
                                resolve: @escaping Resolve,
                                reject: @escaping Reject)

    func isDeviceConnected(_ deviceIdentifier: String,
                           resolve: Resolve,
                           reject: Reject)

    func discoverAllServicesAndCharacteristicsForDevice(_ deviceIdentifier: String,
                                                        transactionId: String,
                                                        resolve: @escaping Resolve,
                                                        reject: @escaping Reject)

    func servicesForDevice(_ deviceIdentifier: String,
                           resolve: Resolve,
                           reject: Reject)

    func characteristicsForDevice(_ deviceIdentifier: String,
                                  serviceUUID: String,
                                  resolve: Resolve,
                                  reject: Reject)

    func characteristicsForService(_ serviceIdentifier: Double,
                                   resolve: Resolve,
                                   reject: Reject)
    
    func descriptorsForDevice(_ deviceIdentifier: String,
                              serviceUUID: String,
                              characteristicUUID: String,
                              resolve: Resolve,
                              reject: Reject)
                              
    func descriptorsForService(_ serviceIdentifier: Double,
                               characteristicUUID: String,
                               resolve: Resolve,
                               reject: Reject)
    
    func descriptorsForCharacteristic(_ characteristicIdentifier: Double,
                                      resolve: Resolve,
                                      reject: Reject)
    

    func readCharacteristicForDevice(_ deviceIdentifier: String,
                                     serviceUUID: String,
                                     characteristicUUID: String,
                                     transactionId: String,
                                     resolve: @escaping Resolve,
                                     reject: @escaping Reject)

    func readCharacteristicForService(_ serviceIdentifier: Double,
                                      characteristicUUID: String,
                                      transactionId: String,
                                      resolve: @escaping Resolve,
                                      reject: @escaping Reject)

    func readCharacteristic(_ characteristicIdentifier: Double,
                            transactionId: String,
                            resolve: @escaping Resolve,
                            reject: @escaping Reject)

    func writeCharacteristicForDevice(_ deviceIdentifier: String,
                                      serviceUUID: String,
                                      characteristicUUID: String,
                                      valueBase64: String,
                                      response: Bool,
                                      transactionId: String,
                                      resolve: @escaping Resolve,
                                      reject: @escaping Reject)

    func writeCharacteristicForService(_ serviceIdentifier: Double,
                                       characteristicUUID: String,
                                       valueBase64: String,
                                       response: Bool,
                                       transactionId: String,
                                       resolve: @escaping Resolve,
                                       reject: @escaping Reject)

    func writeCharacteristic(_ characteristicIdentifier: Double,
                             valueBase64: String,
                             response: Bool,
                             transactionId: String,
                             resolve: @escaping Resolve,
                             reject: @escaping Reject)

    func monitorCharacteristicForDevice(_ deviceIdentifier: String,
                                        serviceUUID: String,
                                        characteristicUUID: String,
                                        transactionId: String,
                                        resolve: @escaping Resolve,
                                        reject: @escaping Reject)

    func monitorCharacteristicForService(_ serviceIdentifier: Double,
                                         characteristicUUID: String,
                                         transactionId: String,
                                         resolve: @escaping Resolve,
                                         reject: @escaping Reject)

    func monitorCharacteristic(_ characteristicIdentifier: Double,
                               transactionId: String,
                               resolve: @escaping Resolve,
                               reject: @escaping Reject)
    
    func readDescriptorForDevice(_ deviceIdentifier: String,
                                 serviceUUID: String,
                                 characteristicUUID: String,
                                 descriptorUUID: String,
                                 transactionId: String,
                                 resolve: @escaping Resolve,
                                 reject: @escaping Reject)
    
    func readDescriptorForService(_ serviceId: Double,
                                  characteristicUUID: String,
                                  descriptorUUID: String,
                                  transactionId: String,
                                  resolve: @escaping Resolve,
                                  reject: @escaping Reject)
    
    func readDescriptorForCharacteristic(_ characteristicID: Double,
                                         descriptorUUID: String,
                                         transactionId: String,
                                         resolve: @escaping Resolve,
                                         reject: @escaping Reject)
    
    func readDescriptor(_ descriptorID: Double,
                        transactionId: String,
                        resolve: @escaping Resolve,
                        reject: @escaping Reject)
    
    func writeDescriptorForDevice(_ deviceIdentifier: String,
                                  serviceUUID: String,
                                  characteristicUUID: String,
                                  descriptorUUID: String,
                                  valueBase64: String,
                                  transactionId: String,
                                  resolve: @escaping Resolve,
                                  reject: @escaping Reject)
    
    func writeDescriptorForService(_ serviceID: Double,
                                   characteristicUUID: String,
                                   descriptorUUID: String,
                                   valueBase64: String,
                                   transactionId: String,
                                   resolve: @escaping Resolve,
                                   reject: @escaping Reject)
    
    func writeDescriptorForCharacteristic(_ characteristicID: Double,
                                          descriptorUUID: String,
                                          valueBase64: String,
                                          transactionId: String,
                                          resolve: @escaping Resolve,
                                          reject: @escaping Reject)
    
    func writeDescriptor(_ descriptorID: Double,
                         valueBase64: String,
                         transactionId: String,
                         resolve: @escaping Resolve,
                         reject: @escaping Reject)
}

extension BleClientManager: BleAdapter { }

File: android/src/main/java/com/bleplx/adapter/BleAdapter.java
package com.bleplx.adapter;

import com.bleplx.adapter.errors.BleError;

import java.util.List;

public interface BleAdapter {

  void createClient(String restoreStateIdentifier,
                    OnEventCallback<String> onAdapterStateChangeCallback,
                    OnEventCallback<Integer> onStateRestored);

  void destroyClient();

  void enable(
    String transactionId,
    OnSuccessCallback<Void> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void disable(
    String transactionId,
    OnSuccessCallback<Void> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  String getCurrentState();

  void startDeviceScan(
    String[] filteredUUIDs,
    int scanMode,
    int callbackType,
    boolean legacyScan,
    OnEventCallback<ScanResult> onEventCallback,
    OnErrorCallback onErrorCallback);

  void stopDeviceScan();

  void requestConnectionPriorityForDevice(
    String deviceIdentifier,
    int connectionPriority,
    String transactionId,
    OnSuccessCallback<Device> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void readRSSIForDevice(
    String deviceIdentifier,
    String transactionId,
    OnSuccessCallback<Device> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void requestMTUForDevice(
    String deviceIdentifier,
    int mtu,
    String transactionId,
    OnSuccessCallback<Device> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void getKnownDevices(
    String[] deviceIdentifiers,
    OnSuccessCallback<Device[]> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void getConnectedDevices(
    String[] serviceUUIDs,
    OnSuccessCallback<Device[]> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void connectToDevice(
    String deviceIdentifier,
    ConnectionOptions connectionOptions,
    OnSuccessCallback<Device> onSuccessCallback,
    OnEventCallback<ConnectionState> onConnectionStateChangedCallback,
    OnErrorCallback onErrorCallback);

  void cancelDeviceConnection(
    String deviceIdentifier,
    OnSuccessCallback<Device> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void isDeviceConnected(
    String deviceIdentifier,
    OnSuccessCallback<Boolean> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void discoverAllServicesAndCharacteristicsForDevice(
    String deviceIdentifier,
    String transactionId,
    OnSuccessCallback<Device> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  List<Service> getServicesForDevice(
    String deviceIdentifier) throws BleError;

  List<Characteristic> getCharacteristicsForDevice(
    String deviceIdentifier,
    String serviceUUID) throws BleError;

  List<Characteristic> getCharacteristicsForService(
    int serviceIdentifier) throws BleError;

  List<Descriptor> descriptorsForDevice(
    String deviceIdentifier,
    String serviceUUID,
    String characteristicUUID) throws BleError;

  List<Descriptor> descriptorsForService(
    int serviceIdentifier,
    String characteristicUUID) throws BleError;

  List<Descriptor> descriptorsForCharacteristic(
    int characteristicIdentifier) throws BleError;


  void readCharacteristicForDevice(
    String deviceIdentifier,
    String serviceUUID,
    String characteristicUUID,
    String transactionId,
    OnSuccessCallback<Characteristic> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void readCharacteristicForService(
    int serviceIdentifier,
    String characteristicUUID,
    String transactionId,
    OnSuccessCallback<Characteristic> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void readCharacteristic(
    int characteristicIdentifer,
    String transactionId,
    OnSuccessCallback<Characteristic> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void writeCharacteristicForDevice(
    String deviceIdentifier,
    String serviceUUID,
    String characteristicUUID,
    String valueBase64,
    boolean withResponse,
    String transactionId,
    OnSuccessCallback<Characteristic> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void writeCharacteristicForService(
    int serviceIdentifier,
    String characteristicUUID,
    String valueBase64,
    boolean withResponse,
    String transactionId,
    OnSuccessCallback<Characteristic> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void writeCharacteristic(
    int characteristicIdentifier,
    String valueBase64,
    boolean withResponse,
    String transactionId,
    OnSuccessCallback<Characteristic> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void monitorCharacteristicForDevice(
    String deviceIdentifier,
    String serviceUUID,
    String characteristicUUID,
    String transactionId,
    OnEventCallback<Characteristic> onEventCallback,
    OnErrorCallback onErrorCallback);

  void monitorCharacteristicForService(
    int serviceIdentifier,
    String characteristicUUID,
    String transactionId,
    OnEventCallback<Characteristic> onEventCallback,
    OnErrorCallback onErrorCallback);

  void monitorCharacteristic(
    int characteristicIdentifier,
    String transactionId,
    OnEventCallback<Characteristic> onEventCallback,
    OnErrorCallback onErrorCallback);

  void readDescriptorForDevice(
    final String deviceId,
    final String serviceUUID,
    final String characteristicUUID,
    final String descriptorUUID,
    final String transactionId,
    OnSuccessCallback<Descriptor> successCallback,
    OnErrorCallback errorCallback);

  void readDescriptorForService(
    final int serviceIdentifier,
    final String characteristicUUID,
    final String descriptorUUID,
    final String transactionId,
    OnSuccessCallback<Descriptor> successCallback,
    OnErrorCallback errorCallback);

  void readDescriptorForCharacteristic(
    final int characteristicIdentifier,
    final String descriptorUUID,
    final String transactionId,
    OnSuccessCallback<Descriptor> successCallback,
    OnErrorCallback errorCallback);

  void readDescriptor(
    final int descriptorIdentifier,
    final String transactionId,
    OnSuccessCallback<Descriptor> onSuccessCallback,
    OnErrorCallback onErrorCallback);

  void writeDescriptorForDevice(
    final String deviceId,
    final String serviceUUID,
    final String characteristicUUID,
    final String descriptorUUID,
    final String valueBase64,
    final String transactionId,
    OnSuccessCallback<Descriptor> successCallback,
    OnErrorCallback errorCallback);

  void writeDescriptorForService(
    final int serviceIdentifier,
    final String characteristicUUID,
    final String descriptorUUID,
    final String valueBase64,
    final String transactionId,
    OnSuccessCallback<Descriptor> successCallback,
    OnErrorCallback errorCallback);

  void writeDescriptorForCharacteristic(
    final int characteristicIdentifier,
    final String descriptorUUID,
    final String valueBase64,
    final String transactionId,
    OnSuccessCallback<Descriptor> successCallback,
    OnErrorCallback errorCallback);

  void writeDescriptor(
    final int descriptorIdentifier,
    final String valueBase64,
    final String transactionId,
    OnSuccessCallback<Descriptor> successCallback,
    OnErrorCallback errorCallback);

  void cancelTransaction(String transactionId);

  void setLogLevel(String logLevel);

  String getLogLevel();
}
